File: frontend/src/services/syllabus.ts

export type SyllabusStatus =
    | 'DRAFT'
    | 'SUBMITTED'
    | 'HOD_APPROVED'
    | 'AA_APPROVED'
    | 'PRINCIPAL_APPROVED'
    | 'PUBLISHED'
    | 'REQUESTEDIT'
    | 'REJECTED'

export interface Course {
    id: number
    code?: string
    name?: string
}

export interface User {
    id: number
    username?: string
    fullName?: string
}

export interface Syllabus {
    id: number
    title: string
    description?: string
    academicYear?: string
    semester?: string
    status: SyllabusStatus
    version: number
    editNote?: string | null

    aiSummary?: string | null
    keywords?: string | null

    course?: Course
    createdBy?: User

    createdAt?: string;
    updatedAt?: string;
}

export interface CreateSyllabusRequest {
    courseId: number
    title: string
    description?: string
    academicYear?: string
    semester?: string

    aiSummary?: string;
    keywords?: string;

}

export interface NoteRequest {
    note: string
}

export interface SyllabusHistory {
    id: number
    syllabusId: number
    title: string
    description?: string
    academicYear?: string
    semester?: string
    status: SyllabusStatus
    version: number
    updatedAt?: string
    // ✅ backend hay trả object user, không phải string
    updatedBy?: User | null;
}

export type Notification = {
    id: number;
    message: string;
    createdAt?: string;
    isRead?: boolean; // backend boolean
};

File: LTJava/src/main/java/com/example/LTJava/syllabus/entity/Syllabus.java
package com.example.LTJava.syllabus.entity;

import java.time.LocalDateTime;

import com.example.LTJava.user.entity.User;

import jakarta.persistence.Column;
import jakarta.persistence.Entity;
import jakarta.persistence.EnumType;
import jakarta.persistence.Enumerated;
import jakarta.persistence.GeneratedValue;
import jakarta.persistence.GenerationType;
import jakarta.persistence.Id;
import jakarta.persistence.JoinColumn;
import jakarta.persistence.ManyToOne;
import jakarta.persistence.PrePersist;
import jakarta.persistence.PreUpdate;
import jakarta.persistence.Table;

@Entity
@Table(name = "syllabus")
public class Syllabus {

    @Id
    @GeneratedValue(strategy = GenerationType.IDENTITY)
    private Long id;

    // Mỗi syllabus thuộc 1 course
    @ManyToOne(optional = false)
    @JoinColumn(name = "course_id")
    private Course course;

    @Column(nullable = false, length = 255)
    private String title;              // tên giáo trình / đề cương

    @Column(length = 4000)
    private String description;        // mô tả ngắn

    @Column(length = 20)
    private String academicYear;       // 2025-2026

    @Column(length = 20)
    private String semester;           // HK1, HK2,...

    @Enumerated(EnumType.STRING)
    @Column(nullable = false, length = 20)
    private SyllabusStatus status = SyllabusStatus.DRAFT;

    private Integer version = 1;       // v1, v2,... (sau này tăng)

    @ManyToOne
    @JoinColumn(name = "created_by")
    private User createdBy;            // giảng viên tạo

    private LocalDateTime createdAt;

    private LocalDateTime updatedAt;

    public Syllabus() {
    }

    @PrePersist
    public void onCreate() {
        this.createdAt = LocalDateTime.now();
        this.updatedAt = LocalDateTime.now();
    }

    @PreUpdate
    public void onUpdate() {
        this.updatedAt = LocalDateTime.now();
    }

    @Column(columnDefinition = "TEXT")
    private String editNote; // nội dung yêu cầu chỉnh sửa

    @Column(columnDefinition = "TEXT")
    private String aiSummary;

    @Column(columnDefinition = "TEXT")
    private String keywords;




    // getters & setters


    public String getAiSummary() {
        return aiSummary;
    }

    public void setAiSummary(String aiSummary) {
        this.aiSummary = aiSummary;
    }

    public String getKeywords() {
        return keywords;
    }

    public void setKeywords(String keywords) {
        this.keywords = keywords;
    }

    public Long getId() {
        return id;
    }

    public void setId(Long id) {
        this.id = id;
    }

    public Course getCourse() {
        return course;
    }

    public void setCourse(Course course) {
        this.course = course;
    }

    public String getTitle() {
        return title;
    }

    public void setTitle(String title) {
        this.title = title;
    }

    public String getDescription() {
        return description;
    }

    public void setDescription(String description) {
        this.description = description;
    }

    public String getAcademicYear() {
        return academicYear;
    }

    public void setAcademicYear(String academicYear) {
        this.academicYear = academicYear;
    }

    public String getSemester() {
        return semester;
    }

    public void setSemester(String semester) {
        this.semester = semester;
    }

    public SyllabusStatus getStatus() {
        return status;
    }

    public void setStatus(SyllabusStatus status) {
        this.status = status;
    }

    public Integer getVersion() {
        return version;
    }

    public void setVersion(Integer version) {
        this.version = version;
    }

    public User getCreatedBy() {
        return createdBy;
    }

    public void setCreatedBy(User createdBy) {
        this.createdBy = createdBy;
    }

    public LocalDateTime getCreatedAt() {
        return createdAt;
    }

    public LocalDateTime getUpdatedAt() {
        return updatedAt;
    }

    public String getEditNote() {
    return editNote;
}

    public void setEditNote(String editNote) {
        this.editNote = editNote;
    }
}
